Composable Pages

dbaines opened the pull request:
I'm making a pull request before this gets too stale and forgotten about. 

### Setup steps 

This has [webpacker](https://github.com/rails/webpacker) integration in to Koi, along with `webpacker-react` for the really nice view helpers. 
This means that you won't be able to run the server with `rails s`, instead you will need to use foreman to run the test app:
  
```
git checkout feature/composable-webpacker
bundle
cd test/dummy
(sometimes you need to bundle here too)
rake db:migrate
yarn install 
foreman start -f Procfile
```
  
### Usage 

Now when loading up a Koi page you will have a composable content block under the WYSIWYG editor.  
Here you can add new components, remove existing components, drag and drop to re-order them using the three-line grabber etc.  
Once you've saved your page, you can view it on the front-end of the test/dummy app to see the output.  

There's no doubt bugs, ux issues, validation issues, configuration requirements that I haven't thought of etc. etc. 
Pull away and throw some use-cases at it! 
  
### RFC 

Here's a link to the RFC for more information, details on developer configuration etc.: 
https://github.com/katalyst/koi/issues/309 
  
### Places needing code review are:  

`composable.rb` - the main concern that gets included in to our models that have composable pages  
`page.rb` - for configuration of composable pages  

**Note**: This is based off of `master` so it's Koi 3 + composable pages.   

**This is likely to become a very important feature of Koi, so I need to make sure you've all looked over it and understand how it works and how to configure it, because you will no doubt be asked to make configuration changes to it in future apps.**
This is a big change so I'd like you to all be onboard.

dbaines commented:
I haven't considered bootstrapping yet. 

At the moment I'm thinking the composable files and the webpacker config will be generated in to the app and it's up to the developer to compile the JS files using webpacker. 
This will allow developers to customise the composable feature without having to worry about overriding files in the Koi gem, and will also allow developers that don't want the composable feature to just ignore it. 

Happy to hear opinions.

Obversity commented:
@dbaines Bug: I created a composable page with no super heroes in my database, added a hero component (but obviously couldn't select one), saved it (it successfully saved), then when I added a super hero  to the database and came back and tried to select it, it didn't seem to save the component with the super hero ID—or at least the front-end page was still failing to load, saying it couldn't find a super hero without an ID.

Obversity commented:
@dbaines Nice work man! This is looking great. I'm slowly working through it, trying to understand how it works.

Question: Could you explain the "Section" thing? Is this supposed to be some kind of nested thing, where you have sections that can each have a list of components to render, so you can specify wether those things are to be rendered full width or in the body content? (It doesn't seem to be functioning if so, but that's how the eventual json data seems to be structured.)

Obversity commented:
I've noticed that it doesn't work in `type :inline` (i.e. nested repeater) sections. It renders the div for the react component but nothing shows up on the interface. 

It'd be pretty edge case where we'd want composable content inside a nested thing like this, so it's probably not worth worrying about. If a nested thing has enough content to warrant composable content, we'll probably be managing those nested models in a nested route/controller rather than in the form itself.

Obversity commented:
I'm thinking we might want to have a few built-in, pre-generated composable components somewhere, for easy reference and to keep things DRY and consistent? 

I.e. in a model you're implementing `Composable` in, you might end up with a `self.composable_field_types` that looks like this:

```ruby
  def self.composable_field_types
    [
      ComposableType::Heading,
      ComposableType::Text,
      ComposableType::WSYIWYG,
      ComposableType::KitchenSink,
      {
         # custom type goes here
      },
      {
          # Aaaand another one
      }
    ]
  end
```

And a `ComposableType` module somewhere (maybe an initializer?) that looks like this:

```
module ComposableType
  
  Heading = {
    name: "Heading",
    slug: "heading",
    fields: [{
      label: "Heading Text",
      name: "text",
      type: "string",
      className: "form--medium"
    },{
      label: "Size",
      name: "heading_size",
      type: "select",
      data: ["2", "3", "4", "5", "6"],
      className: "form--auto"
    }]
  }
  
  Text = {
    name: "Text",
    slug: "text",
    fields: [{
      name: "body",
      type: "textarea"
    }]
  }

  ...
end
```

Worth discussing!

jsidoryn commented:
@Obversity @dbaines thanks for keeping this alive after a bit of a break with doing work on koi. i'll try and get it set up locally for a run through but at worst, let's find some time on wed to touch base on it

Obversity commented:
Random UX thoughts: 

The UX of adding composable components doesn't feel quite right, in my opinion, for content creation. It needs to be as easy as possible. At the moment, it's three clicks to add a component, four clicks to start actually typing content. 

We could probably cut this down to one with maybe a scrollable sidebar to the left of the form, with all the component types listed, where you can click one and it adds it to the flow and puts focus on the first field. (Regardless of the solution, I think we should try to cut down the number of clicks a user needs to make to add a component. If we're thinking about using this for general content creation, it needs to be smooth. Definitely worth looking at similar tools to see how they solve it.)

Another thing, it might help to have the concept of 'composable component settings', i.e. fields that you don't appear in the normal flow of fields, that you can click a 'settings' button to view in some other way (maybe a modal?). Stuff like heading size, image size, any random flags (i.e. checkboxes) against things. 

(The more I think about this idea the less I'm sold on it, but I'll have to see what kind of configuration we need on components in a real world project to decide whether it's worth it.

My main thought process was that it'd be nice to just be able to flop open a component and see the primary content of it without being distracted by all the configuration, but at the same time, that makes configuration of that content several more clicks worth of work, and makes it harder for admins to understand why content is behaving in a certain way on the front-end.)

dbaines commented on app/views/shared/_composables.html.erb:
```diff
@@ -0,0 +1,27 @@
+<%- data.each do |section| -%>
```
Yep! That's what I'm thinking. We'll end up generating all the composable stuff in to the project. Especially since this is the front-end view! Wouldn't be very nice having this stuck in Koi.

dbaines commented on config/initializers/simple_form.rb:
```diff
@@ -164,7 +164,7 @@
 
```
I realise this was a comment from potentially a long time ago, but [html5 validations are very well supported](https://caniuse.com/#feat=form-validation). Seeing as this is Koi, which is even more strict about browser support than the front-end I'd say it's pretty safe. 
Even [Ornament now has it turned on by default](https://github.com/katalyst/ornament/blob/master/test/dummy/config/initializers/simple_form.rb#L116). 

Client-side validations are very dumb, all you can match is a regex-like pattern and requiredness. This is not in any way a replacement for rails validations. All I'm planning on using it for is setting requiredness on some fields. 

We may need to think about how to display errors if submitted data fails a backend check which I haven't considered at all. Since it's technically a JSON blob, I'm really not even sure where to begin! I'm happy to hear any ideas on this issue.

dbaines commented on test/dummy/app/javascript/packs/hello_react.jsx:
```diff
@@ -0,0 +1,26 @@
+// Run this example by adding <%= javascript_pack_tag 'hello_react' %> to the head of your layout file,
```
Yep! That was the default react component generated by webpacker, I've removed it now.

dbaines commented:
@Obversity 

Regarding model configuration:

I've created an overridable `composable_components.rb` file that contains the list of all components. There's a default file in Koi that will get generated in to your project, while the test dummy app has it's own that adds things like `hero` and `kitchen_sink`.

I've added a new crud config option like so to define which composable components a model has access to:

```ruby
    config :admin do
      index   fields: [:id, :title]
      form    fields: [:title, :description, :composable_data],
              composable: ["section", "text", "heading", "hero", "kitchen_sink"]
    end
```

This feels like a pretty good configuration technique to me. 

I tried to go with your recommendation of making it a `has_crud composable: true` option rather than a concern but I got a bit stumped when it came to including the concern if that was true. I'm guessing it needs to be in `crud_admin_additions.rb` everything I tried crashed. 

Can we pair on converting it to a `has_crud` option rather than a concern? Or if you're happy to do it yourself I don't mind. That sort of thing doesn't really bother me as it doesn't affect configuration it's really just where the code lives in rails land.

dbaines commented:
@Obversity I think this is in a pretty nice place now. 

### React on Rails vs Webpacker-React

Probably not that relevant to you, but I've switched from [webpacker-react](https://github.com/renchap/webpacker-react) to [react_on_rails](https://github.com/shakacode/react_on_rails), they are mostly the same but react on rails seems to have more community support behind it and development seems much more steady. These gems are what provide the `react_component` view helper and offer nice things like server side rendering that we can opt in to.

### Generating / Bootstrapping updates

I've added the composable js files in to the application template and the shared view partials to make kick starting a composable app pretty seamless. 

**Note**: As I mentioned offline, it does require Ornament, specifically the `develop` branch right now. 
The develop branch is the one that sets up webpacker, react on rails and configures everything. Because of this I have the develop branch specified in the app template in this branch. 
When the webpacker config gets pushed to master in Ornament, we will then be safe to change the app template back to pull from Ornament master. I'm hoping to get this done within the next couple of weeks.

### Todos

There are a couple of things I couldn't figure out, if you're happy to take a bash when you get some time go for it:

1. My previous comment regarding the `has_crud` config. Currently composable is a concern, but if you could reconfigure it so it can be `has_crud composable: true` instead that would be great. 
1. Koi has a default "library" of components in `lib/koi/composable_content.rb`, which the client app can override in eg. `app/config/initializers/koi/composable_content.rb` - This works really well but on starting the app we get a warning that the constant is defined twice. Is this a problem? If so can you think of a better way to manage this? Or should we just ditch the koi one since it will always get generated anyway?
1. (Possibly more of an Ornament change) React On Rails generates a route "hello_world", if you can figure out a way to remove this route after it's added that would be great. It's not a huge problem just noise. I've already removed the controller and views that it generates, just that pesky route that's left over.
